lua/plenary/path.lua

```diff
@@ -515,30 +515,73 @@ function Path:rmdir()
   uv.fs_rmdir(self:absolute())
 end
 
+---Rename this file or directory to the given target, and return a new Path instance pointing to target. On Unix, if target exists and is a file, it will be replaced silently if the user has permission. On Windows, if target exists, FileExistsError will be raised. target can be either a string or another path object:
+---@generic T: Path
+---@param opts { new_name: Path|string }
+---@return T
 function Path:rename(opts)
+  -- TODO: For reference, Python's `Path.rename()` actually says/does this:
+  --
+  -- > On Unix, if target exists and is a file, it will be replaced silently
+  -- > if the user has permission.
+  -- >
+  -- > On Windows, if target exists, FileExistsError will be raised. target
+  -- > can be either a string or another path object.
+  --
+  -- The behavior here may differ, as an error will be thrown regardless.
+
+  local self_lstat, new_lstat, status, errmsg
   opts = opts or {}
-  if not opts.new_name or opts.new_name == "" then
-    error "Please provide the new name!"
-  end
+  assert(opts.new_name and opts.new_name ~= "", "Please provide the new name!")
+  self_lstat, errmsg = uv.fs_lstat(self.filename)
+
+  -- Cannot rename a non-existing path (lstat is needed here, `Path:exists()`
+  -- uses stat)
+  assert(self_lstat, ("%s: %s"):format(errmsg, self.filename))
 
+  -- BUG
   -- handles `.`, `..`, `./`, and `../`
   if opts.new_name:match "^%.%.?/?\\?.+" then
     opts.new_name = {
       uv.fs_realpath(opts.new_name:sub(1, 3)),
-      opts.new_name:sub(4, #opts.new_name),
+      opts.new_name:sub(4),
     }
   end
 
   local new_path = Path:new(opts.new_name)
-
-  if new_path:exists() then
```

Comment:
This was a bug because `Path:exists()` uses stat, not lstat. `new_name` could be a bad symlink, in which case it would incorrectly think that the file doesn't exist (rename(2) does not resolve the final path component if it's a symlink).